packages/application/src/frontend.ts

```diff
@@ -0,0 +1,213 @@
+   *
+   * @returns an HTMLElement or undefined, if none is found.
+   */
+  contextMenuFirst(
+    test: (node: HTMLElement) => boolean
+  ): HTMLElement | undefined {
+    for (let node of this._getContextMenuNodes()) {
+      if (test(node)) {
+        return node;
+      }
+    }
+    return undefined;
+  }
+
+  /**
+   * A method invoked on a document `'contextmenu'` event.
+   */
+  protected evtContextMenu(event: MouseEvent): void {
+    this._contextMenuEvent = event;
+    super.evtContextMenu(event);
+  }
+
+  /**
+   * Gets the hierarchy of html nodes that was under the cursor
+   * when the most recent contextmenu event was issued
+   */
+  private _getContextMenuNodes(): HTMLElement[] {
+    if (!this._contextMenuEvent) {
+      return [];
+    }
+
+    // This one-liner doesn't work, but should at some point in the future
+    // `return this._contextMenuEvent.composedPath() as HTMLElement[];`
+    // cf. (https://developer.mozilla.org/en-US/docs/Web/API/Event)
+
+    const nodes: HTMLElement[] = [this._contextMenuEvent.target as HTMLElement];
+    while (
+      'parentNode' in nodes[nodes.length - 1] &&
+      nodes[nodes.length - 1].parentNode &&
+      nodes[nodes.length - 1] !== nodes[nodes.length - 1].parentNode
+    ) {
+      nodes.push(nodes[nodes.length - 1].parentNode as HTMLElement);
+    }
+    return nodes;
+  }
+
+  private _contextMenuEvent: MouseEvent;
+}
+
+/**
+ * The namespace for `JupyterFrontEnd` class statics.
+ */
+export namespace JupyterFrontEnd {
+  /**
+   * The options used to initialize a JupyterFrontEnd.
+   */
+  export interface IOptions<T extends Shell = Shell, U = any>
+    extends Application.IOptions<T> {
+    /**
+     * The document registry instance used by the application.
+     */
+    docRegistry?: DocumentRegistry;
+
+    /**
+     * The command linker used by the application.
+     */
+    commandLinker?: CommandLinker;
+
+    /**
+     * The service manager used by the application.
+     */
+    serviceManager?: ServiceManager;
+
+    /**
+     * Promise that resolves when state is first restored, returning layout
+     * description.
+     */
+    restored?: Promise<U>;
+  }
+
+  /**
+   * A minimal shell type for Jupyter front-end applications.
+   */
+  export type Shell = Widget & {
+    /**
+     * Activates a widget inside the application shell.
+     *
+     * @param id - The ID of the widget being activated.
+     */
+    activateById(id: string): void;
+
+    /**
+     * Add a widget to the application shell.
+     *
+     * @param widget - The widget being added.
+     *
+     * @param area - Optional region in the shell into which the widget should
+     * be added.
+     *
+     * @param options - Optional flags the shell might use when opening the
+     * widget, as defined in the `DocumentRegistry`.
+     */
+    add(
+      widget: Widget,
+      area?: string,
+      options?: DocumentRegistry.IOpenOptions
```

Comment:
This still makes me a bit nervous. The semantics for how to do layout in a 1D tab panel and a 2D dockpanel are pretty different, and it seems kind of confusing to the developer to handle both with the `DocumentRegistry.IOpenOptions`.

Reply:
This was my best attempt to abstract shell functionality in a way that would accommodate all the shells I could think of; but I'm open to suggestions.

Reply:
My suggestion below is that we move these options from the `DocumentRegistry` to the `ILabShell` namespace.